KLR/NKI/Types.lean

```diff
@@ -0,0 +1,137 @@
+-/
+inductive SNat (nnat : Nat)
+  | param (idx : Fin nnat)
+  | const (n : Nat)
+
+/--
+The universe of types.
+
+`nnat` is the total number of universal quantifiers in scope for `SNat`s.
+
+`ntyp` is the total number of universal quantifiers in scope for `STyp`s.
+
+Example:
+`∀ n m n, ∀ t1 t2, tensor[t1, 10, m]` ≡ `(STyp.tensor [SNat.const 10, SNat.param 1] (STyp.param 1) : STyp 3 2)`
+-/
+inductive STyp (nnat ntyp : Nat)
+  | param (idx : Fin ntyp)
+  | none
+  | bool
+  | int
+  | float
+  | string
+  | tuple (typs : List (STyp nnat ntyp))
+  | tensor (shape : List (SNat nnat)) (dtype : Core.Dtype)
+  | func (dom ran : STyp nnat ntyp)
+
+namespace NominalTypes
+  /-!
+  # An alternative types definition using nominal variable binding
+
+  This requires a separate well-formedness proof to show that varaible references are valid.
+  -/
+
+  inductive Srt
+    | snat
+    | styp
+
+  structure TypParam where
+    name : String
+    sort : Srt
+
+  inductive SNat
+    | param (name : String)
+    | const (n : Nat)
+
+  inductive STyp
+    | param (name : String)
+    | none
+    | bool
+    | int
+    | float
+    | string
+    | tuple (typs : List STyp)
+    | tensor (shape : List SNat) (dtype : Core.Dtype)
+    | func (dom ran : STyp)
+
+  inductive Typ'
+    | snat (sn : SNat)
+    | styp (st : STyp)
+
+  /--
+  TODO: what about name collision in `params`?
+  -/
+  structure Typ where
+    params : List TypParam
+    typ : Typ'
+
+  inductive SNat.WellFormed : List TypParam → SNat → Prop
+    | param {ps : List TypParam} {name : String}
+        : ⟨name, .snat⟩ ∈ ps → SNat.WellFormed ps (.param name)
+    | const {ps : List TypParam} {n : Nat}
+        : SNat.WellFormed ps (.const n)
+
+  inductive STyp.WellFormed (ps : List TypParam) : STyp → Prop
+    | param : ⟨name, .styp⟩ ∈ ps → STyp.WellFormed ps (.param name)
+    | none : STyp.WellFormed ps .none
+    | bool : STyp.WellFormed ps .bool
+    | int : STyp.WellFormed ps .int
+    | float : STyp.WellFormed ps .float
+    | string : STyp.WellFormed ps .string
+    | tuple : (∀ ty ∈ typs, STyp.WellFormed ps ty) → STyp.WellFormed ps (.tuple typs)
+    | tensor : (∀ s ∈ shape, s.WellFormed ps) → STyp.WellFormed ps (.tensor shape dtype)
+    | func : STyp.WellFormed ps dom → STyp.WellFormed ps ran → STyp.WellFormed ps (.func dom ran)
+
+  inductive Typ'.WellFormed : List TypParam → Typ' → Prop
+    | snat : sn.WellFormed ps → Typ'.WellFormed ps (.snat sn)
+    | styp : st.WellFormed ps → Typ'.WellFormed ps (.styp st)
+
+  def Typ.WellFormed (t : Typ) : Prop :=
+    t.typ.WellFormed t.params
+
+end NominalTypes
+
+end KLR.NKI
+
+namespace TypesExamples
+
+def List.swap (l : List α) (i₁ i₂ : Nat) : List α :=
+  if h : i₁ < l.length ∧ i₂ < l.length then
+    (l.set i₁ l[i₂]).set i₂ l[i₁]
+  else
+    l
+
+open KLR NKI
+
+def transpose {nnat ntyp : Nat} {dtype : Core.Dtype} {shapes : List (SNat nnat)} : STyp nnat ntyp :=
+  .func (.tensor shapes dtype) (.tensor (List.swap shapes 1 3 ) dtype)
```

Comment:
Nice example. So for sth like matmul2D we'd have this?
```
matmul2D : .func2 (.tensor [l, x] dtype) (.tensor [x, r] dtype) (.tensor [l, r] dtype)
```

Reply:
Btw, this is `matmul2d` in the nominal representation:
```lean
-- `∀ {s1 s2 s3} {t}, tensor[t, s1, s2] → tensor[t, s2, s3] → tensor[t, s1, s3]`
def nmatmul2d {dtype : Core.Dtype} : Typ :=
  ⟨
    [⟨"s1", .snat⟩, ⟨"s2", .snat⟩, ⟨"s3", .snat⟩, ⟨"t", .styp⟩],
    .styp <|
      .func (.tensor [.param "s1", .param "s2"] dtype)
        <| .func (.tensor [.param "s2", .param "s3"] dtype) (.tensor [.param "s1", .param "s1"] dtype)
  ⟩
```

Reply:
What is the `t` in the comment, e.g. `tensor[t, s1, s2]`?

Reply:
1. I guess we should use the Python convention of tuple-ing the arguments. Assuming one important reason for the type checker is to help NKI authors, they will be writing Python and currying is not super common imo. 
2. I'd thought about this in the context of allowing a function to take "castable" dtypes, e.g. adding Int32 and Int64 tensors without an explicit `astype`. I gave up on that when I figured we could have the compiler insert `astype` calls either at the array level or when indexing over the two tensors. But still, it's worth thinking about the pros and cons.